### DISCUSSION TITLE
Request by poly

### DISCUSSION THREAD

**Discussion by ogamespec** [emu-russia/dmgcpu#97] [General]
Hi,

@msinger, moved the topic into a separate discussion (from #13).

😃 

## RST Pad

![RST_Fused](https://user-images.githubusercontent.com/5828819/194948894-d95e3572-efe7-4e85-9047-97f9b7f01db5.jpg)

## XI Pad

![XI_Fused](https://user-images.githubusercontent.com/5828819/194948907-473648f4-431d-4512-9140-ee4fa24956ab.jpg)

## SIN Pad

![SIN_Fused](https://user-images.githubusercontent.com/5828819/194948898-8700e70e-8a70-45de-917d-f3080e288077.jpg)

## SCK Pad

![SCK_Fused](https://user-images.githubusercontent.com/5828819/194948895-3595e168-3abf-4a71-b8fc-46c7b4d58b71.jpg)

## Stacked Nots

![StackedNots_1](https://user-images.githubusercontent.com/5828819/194948905-483991e7-6882-4f2a-8c5d-f031fa44be12.jpg)
![StackedNots_2](https://user-images.githubusercontent.com/5828819/194948906-716999d0-b016-406e-8917-842395319726.jpg)

## Spare

![Spare_Fused](https://user-images.githubusercontent.com/5828819/194948902-745fa436-40b3-4e74-a1eb-941a3915635f.jpg)

## Source pics (slides)

[Slides_1.zip](https://github.com/emu-russia/dmgcpu/files/9749169/Slides_1.zip)

**Comment by ogamespec** [comment]
I don't know if any of this helps or not, but I see that the first Stacked Nots (below the BootROM) have gone to the poly and there is no M1 anymore :(

But it seems to me they have nowhere to go but to join together one after another in a train. 😃

**Reply by msinger** [reply]
Oh thank you! This helps a lot.

Now the connection from the huge input pads can be clearly seen, how it is winding up and down under the metal.

For the standard cells, the active areas and n-wells have never been so perfectly visible before. Also the connections between M1 and M2 can be seen as rectangles. The other die shots we had were so blurry that those connections looked circular.

Oh, and 0018 is a perfect shot of those big DFFSR flip flops. This will save me a huge amount of time, seeing poly and active so clearly. And those long strips of small vias under the supply rails are really nothing more than well connections like I guessed.

I'm still not getting what those spare cells are supposed to be (0017). But I think I will be able to figure out their layout now.

**Comment by ogamespec** [comment]
Have you seen a ripped off M2 on a silicon-pron?

There's a DMG CPU-B and M1 for the parts you asked for there you can clearly see. In particular I see that for Stacked Nots1 (under BootROM) the outputs do not go anywhere (floating).

https://siliconpr0n.org/map/nintendo/dmg-cpu-b/s1-1_mit20x/#x=13088&y=6496&z=8

**Reply by msinger** [reply]
Yes, I've seen that. That's where I got my pron from. :)
You can switch to the blue-ish die shot in the menu in the top right corner of my browser map.
I found one connection between those cells on M2:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line[]&cells=0&labels=0&line[0]=-159.64,143.89,-159.64,144.31,-159.94,144.31,-159.94,143.89,-159.64,143.89
But I can see nothing on the other cells. I guess they are connected directly at those big M1-M2 vias. Because of the blurriness of the shot it can't be seen, maybe. But they have to be connected somehow. Otherwise the boot ROM and the RAMs wouldn't work. On your pictures, it was clear that there is no poly or active connection, so it must be M2 at the vias. I can't see another option.

**Comment by msinger** [comment]
Hi, just wanted to let you know, I have finally drawn the layout for the RST pad from your picture: http://iceboy.a-singer.de/doc/dmg_cells.html#ibuf_a
Sorry that I had this laying around for so long, after you spent extra time for making these images. I nearly finished the boot ROM as well. I still have to transfer some parts from paper to Electric VLSI. Getting the scaling correct in relation to the other cells is always tricky.
I hope it's okay for you that I've added the RST pad picture to that page. I also added a link at the very top.

**Reply by ogamespec** [reply]
Cool, always wondered where the protection diodes were and how they were formed, that's useful information!
About the pictures - no problem, the license for this repository is CC 1.0 aka "Do what you want I don't care" :)
If there are any other areas you need to take pictures of, let me know. The chip is still at home somewhere :)

**Reply by msinger** [reply]
I'm happy you like the diodes! :)
That's good to know, that the chip is still around.

Régis and I, we would very much like to add the analog sound stuff to the schematic. It would be great to have better pictures of those. I marked some areas, but of course those analog things exist two or four times, so there are other options.
CH4 amp and volume select, plus head of mixer resistors CH1 and VIN:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line[]&cells=0&wires=0&line[0]=-168.81,164.84,-175.44,164.59,-175.78,169.94,-187.28,170.59,-187.19,175.72,-168.75,175.69,-168.81,164.88
Voltage divider (low prio; already good visibility):
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-122.28,167.75,-126.47,167.78,-126.41,172.38,-124.91,176.78,-122.66,176.94,-122.28,167.78
Output amp, final volume of left channel:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-214.94,151.38,-228.34,151.75,-228.72,173.16,-220.00,173.34,-219.38,171.72,-208.88,171.06,-208.88,163.53,-214.81,162.50,-214.94,151.41
And of course we would need the analog output and input pads:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-236.16,206.72,-253.59,206.69,-253.41,228.78,-236.19,229.22,-236.13,206.69

While making the layout of the ~RST pad, I heavily relied on your picture for those diodes and the n-wells. Learning from this, I think it would be very helpful to have pictures of the other IO pads as well. The more metal you get off the better. But I understand it may be risky taking off to much.
A simple output cell can be located right next to the VDD pad:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-236.34,178.31,-253.47,178.00,-253.44,189.97,-236.50,190.13,-236.34,178.31
The second type of output cell would be right next to the analog pads and another one of the same type would be just around the corner:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-235.69,226.88,-253.69,226.94,-253.50,238.69,-226.19,237.63,-226.19,254.56,-237.19,254.81,-235.63,226.88
You already did the SIN pad. I don't understand the part near the bottom left, what looks like a transistor maybe:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-168.08,240.14,-171.50,240.16,-171.61,245.27,-168.23,245.08,-168.11,240.13
If it is not possible to get more metal of from that area without destroying it completely, there are plenty of other opportunities of the same type of cell. For example, one of those four here:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-183.75,239.38,-183.63,255.13,-226.13,255.25,-226.00,239.13,-183.63,239.50
Or you could use any of the D or MD pads in the top left corner.
There is this part of the ~NMI cell, that is different than it is in the ~RST pad that you already took a picture of. I had to guess the shape of the NMOS that is used as pull-up resistor, because of bad visibility. (Low priority)
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-37.73,129.05,-37.55,133.84,-41.86,134.30,-42.25,128.77,-37.72,129.05

What I would also find interesting are those things in each of the four corners, could be diodes or capacitors:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-31.09,7.91,-40.28,8.13,-40.34,17.19,-31.00,17.03,-31.19,7.88

There are some big pads with hundreds of holes. Maybe substrate connections to ground? I don't understand why they are only in this area:
http://iceboy.a-singer.de/dmg_cpu_b_map/?view=line%5B%5D&cells=0&wires=0&line%5B0%5D=-57.81,177.56,-57.81,169.38,-61.44,169.41,-61.44,163.19,-64.09,163.09,-64.34,177.69,-57.81,177.53
I don't think that pictures will reveal anything underneath, so you don't have to bother. I'm just interested what you think this could be? Is it typical to place that around the clock and reset logic maybe?

The different types of flip flops are very densely packet cells. You already took perfect shots of the one with DFF with set&reset inputs. It would be good to have pictures of the other flip flops as well. It's already getting late, I will try to look for some specimen that can be easily located close to the border or near big structures, so that it's easier to find under the microscope.

Yeah, I don't know how much work this is. Don't feel pressured, just do what you can when you are in the mood for it. There is no hurry, I still have a lot of other cells to draw anyway. :)

**Reply by ogamespec** [reply]
> n-wells

Isn't that a typo? The DMG SoC uses p-wells everywhere as far as I can see.

**Reply by msinger** [reply]
No, I double checked. I'm confusing n- and p-well all the time, but then the rules checker of Electric VLSI always sets me straight again. I guess you are the same as me there. :) You mean PMOS transistors need a well, right? Yeah, that's the n-well. It needs opposite doping than the transistors that sit in it.

**Reply by ogamespec** [reply]
Right. That's right. So I need to relabel this stuff everywhere in the chips :)

**Reply by msinger** [reply]
Oh no :worried:

**Comment by ogamespec** [comment]
Analog stuff: https://drive.google.com/drive/u/2/folders/1bMiEECDJHcIFA21s0CERc2Yqiyr9NCi-

Previews:
![DMG_B_Analog_0001_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/5ba03aba-aa3b-4739-8d8a-9501eaeb5d8a)
![DMG_B_Analog_0002_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/5950553d-c25f-4ab7-983d-708d019adf57)
![DMG_B_Analog_0003_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/1abf7a7a-1c6d-464e-b3b6-b452d707ce02)
![DMG_B_Analog_0004_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/7f1905f6-79ec-43a6-98ba-5f6c30eea12b)

**Reply by ogamespec** [reply]
I'll do the rest after the weekend :) 
Look for more specimen cells (DFF), if you make it, I'll grab them on Monday, too :)

**Reply by msinger** [reply]
Thank you. That was so fast. I stared at the analog IOs for a while. I think I have an idea now how they are implemented. If I see it correctly, they are bidirectional. I think they are just a big transmission gate, meaning, one big NMOS and one big PMOS, both being permanently enabled, having their gate connected to VDD and GND respectively. They are a bit different at the top though, I don't know why yet.

**Reply by msinger** [reply]
Nah, only the analog input is a transmission gate. The outputs are different.

**Reply by msinger** [reply]
Okay, I created a secret new floorplan for the map. It can only be enabled by this link:
http://iceboy.a-singer.de/dmg_cpu_b_map/?floorxxx=1&floorplan=0&wires=0&cells=0
It shows the cells in different colors. The flip flops we need are colored in yellow, orange and blue. So we need one specimen of each of those colors. I think a good spot, where all those three colors are close together is directly left of the SOUT and SIN pads on the right hand side of the die shot.
Here a description of all the colors:
* Yellow: DFFR_A
* Orange: DFFR_B2
* Blue: TFFD (the toggle flip flop)
* Red: DFFR_B (just a stripped down version of DFFR_B2, so it is not important)
* White: MUX
* Gray: MUXI
* DarkGray: DR_LATCH
* LightGreen: D_LATCH_A
* DarkGreen: D_LATCH_A2
* Green: D_LATCH_B
* Black: All other cells

I haven't colored the DFFSR flip flop, because you already took perfect pictures of it with the spare cells.
I also color coded the mux and latch cells, which I already layouted, because I didn't clearly see the doped and poly areas and had to do some guess work. So if you like, you can catch a few of those too. But the really important ones are the yellow, orange and blue ones.

I will remove this floorplan if you're done.

**Reply by msinger** [reply]
This is the area I mean:
![image](https://github.com/emu-russia/dmgcpu/assets/186443/cb7e8dba-6241-41cf-abf9-08000d9d78cd)

**Reply by ogamespec** [reply]
I can't say for sure right now where the "skin" peeled off is better and where it's worse. Thanks for the map, as soon as I get to the microscope I will collect you samples of cells, where you can see better poly.

**Comment by ogamespec** [comment]
https://drive.google.com/drive/u/2/folders/1-VlFxGOgzJBf4ymO2J9WbGne3Zc96t0w

S Pad:
![S_Pad_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/ec63cbe3-6836-48c6-b399-f5bcf7665693)

P14+P15+Bottom right cross:
![P14_P15_Fused_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/4e0f6484-1070-4a10-b020-74c01e65be10)

/NMI Pad (gone):
![Pad_nNMI_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/3b34a83f-4c68-4fe4-bc5c-805dc77dd663)

pad comb (somewhere from the top pads):
![Pad_Weird_Comb_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/0beab3f5-10ae-438a-a24f-fbfdba063fbf)

Cross-sections:
bottom left: ![Bottom_Left_Xross_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/008625cf-40f9-4318-a8eb-d1e6b76d815a)
top left: ![Top_Left_Xross_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/8e478b20-0bb5-4200-a824-1c1e633cdb35)

Mass viases section:
![Mass_Vias_Area_sm](https://github.com/emu-russia/dmgcpu/assets/5828819/2c5609a4-1a2d-4560-94b2-8111a6d77054)

**Reply by ogamespec** [reply]
Single slides should be rotated +0.20 deg. Stitched images are aligned.

**Reply by msinger** [reply]
Interesting things. :thinking: 

The XCrosses look the same as the thing that is right next to the SCK and SIN pads for example. I think it's a polarity protection, implemented by a lot of NMOSes, which have their gates wired to GND and their source/drain terminals wired to GND and VDD. When the polarity is correct, then they are not conductive, because the gate is grounded. If the polarity is reversed, then the gate is VDD and they are conductive. But there is more. It is possible that they also implement capacitors. Could be diodes as well, but for diodes I can't see a reason for the comb shapes. (By comb shapes I mean the small ones on the N/P active layer, not the big metal combs.) It's just a wild guess, but I think the comb shape could improve the capacity by not allowing the charges to push themselves too far away from each other, making better use of the area. Similar like braid wire counters the skin effect, allowing more current per area. Does this make sense, or do you have another idea? Could also just be my confirmation bias, because I was suspecting capacitors before.

**Reply by ogamespec** [reply]
I don't know much about analog components of integrated circuits, because I do most of my analysis of logic, and there is enough knowledge that a transistor is a "switch with a short gate memory" 😄 
Most likely these square patterns are additional ESD protection, to dissipate static voltage (guess).

**Reply by ogamespec** [reply]
Although it may be some simple protection against a change of polarity. After all, we have a handheld where the user can easily insert the batteries backwards 🤣

**Comment by ogamespec** [comment]
Cells:

https://drive.google.com/drive/u/2/folders/108xxdoMb1MvxBl9psPZuUqRCLP_3O43-

I hope I got the colors right 😃 

I don't really figure out where the Green/DarkGreen and Gray/DarkGray are. If anything, I'll do a follow-up.

**Reply by ogamespec** [reply]
It would be nice if instead of Furrtek cryptic designations, you could write on the cell its type and just its ordinal number (g0, g1, g2 etc.).

**Reply by msinger** [reply]
Sorry about the green and gray colors, I didn't notice that. You got everything. You just mixed up the gray and dark-gray (MUXI and DR_LATCH). Maybe the gray looks darker than the dark-gray because of the way Leaflet mixes the transparent layers.

The quality of your pictures is really so much better than the others. I will fix the shapes of the poly and active areas of the latches and muxes. It will safe me a lot of time and guess work when I'm layouting the flip-flops. Thank you very much! :kissing_heart:

**Reply by msinger** [reply]
I was thinking about adding the cell type to the label layer of the map, but I haven't done it yet, because I need to implement it in a generic way. I don't want to place the text for each cell type manually. My converter tool must find a good non-overlapping placement on its own. For bigger cells, this isn't a problem, but the small inverters are tricky. When I reduce the font size, then it looks like shit, because the resolution of the pictures is much lower than the maximum zoom level of the map. I can't increase this resolution, because the pictures generated by my tool are already 16Kx16K and this takes multiple gigabytes of RAM on my PC when running it on the Mono Framework. The C# object that I use for drawing needs more memory than just the raw pixel data, because all drawing changes are put in a fifo and get applied delayed somehow to speed things up for smaller images, I don't really know. I call Flush functions very often in a desperate attempt to counter that. If I don't flush, then my kernel runs out of memory and starts killing shit. But yes, having the cell type on the map is also something I really want to have.

Régis and I, we wanted to stay compatible to Furrtek, so we chose to keep the cryptic naming scheme. We already had documents and also discussions in which we are referring to the cells in Furrtek's names. Would you prefer it if the cells were simply numbered, or did you propose that for making more space for the cell type?

**Reply by ogamespec** [reply]
I don't mind leaving the Furrtek names, if that's more convenient for you :)
Suggestion: make a Checkbox that will show the cell types instead of their names.

**Reply by msinger** [reply]
I have to think about how to add the types there. For now I have added more information to the popup that opens when you click on a cell.

**Comment by msinger** [comment]
There is one part of the boot ROM that I don't understand. It is at the left hand side of the boot ROM, where the row selector active area is connected to VDD. My problem is, there is a visible edge around the VDD vias. This makes it look like all of those vias are connected to the same area below. But this can't be true, because this would mean that there are no well connections. In my layout, I made the six vias in the middle the well connection, but I'm not sure if this is correct. Do you have better visuals of that area? Could also be the VDD metal that gets wider around the vias, but it's not really clear.
![image](https://github.com/emu-russia/dmgcpu/assets/186443/ee0de162-77d5-404d-b20c-1063ef964c6a)

**Reply by ogamespec** [reply]
was on vacation, when I get to the microscope will take pictures 😃

**Reply by ogamespec** [reply]
![DMG-BootRom-Area](https://github.com/emu-russia/dmgcpu/assets/5828819/bd9e5eb9-c77a-4819-8b3d-0e11021d941a)
Done.

**Reply by msinger** [reply]
OMG, there is no well connection at all. Electric's rules checker will kill me. 😱
I don't understand how latch-up is prevented here. The NMOS transistors are right next to them. Maybe they forgot. I wonder if the C revision or later Game Boy models have this fixed.

**Comment by ogamespec** [comment]
@msinger 

I've been tidying up my chip collection and decided to save what's left over from my DMG-CPU B polishing sessions.

I'll put it here:

https://drive.google.com/drive/u/3/folders/11EcTeWtmXpVKZfzkdOcWubE6-Ka3VMSu

ClkGen and PPU are already there now.

**Reply by msinger** [reply]
Great! Maybe I can stitch them together for a third map option. The detail of those pictures would be worth it, even if it doesn't cover the whole chip surface. And it's always good to have more pictures of those cells.
Thank you!

**Reply by ogamespec** [reply]
Updated
https://drive.google.com/drive/u/3/folders/11EcTeWtmXpVKZfzkdOcWubE6-Ka3VMSu

Cells1: ClkGen
Cells2: PPU
Cells3: APU
Cells4: IO
Cells5: Arb
MegaCells_HiRAM: HiRAM

Cells are made on a 20x objective, MegaCells are made on a 40x objective.

**Reply by ogamespec** [reply]
Updated
+ MegaCells_BootROM_WaveRAM

![image](https://github.com/user-attachments/assets/1caa75b0-084c-449b-9319-80829ecbfeb3)

**Comment by ogamespec** [comment]
Updated Pads: https://drive.google.com/drive/u/2/folders/1B_YZys5RnkWCs_l3cic8TrlrKJ8l5g5u

More will be added: IOBUF_A, IOBUF_B.

Unfortunately, /NMI (IBUF_B) cannot be recovered :(

**Reply by ogamespec** [reply]
IOBUF_A, IOBUF_B added https://drive.google.com/drive/u/2/folders/1B_YZys5RnkWCs_l3cic8TrlrKJ8l5g5u
(from two locations, with different degrees of polish)

**Reply by msinger** [reply]
They look good, thank you!

**Comment by msinger** [comment]
Hi @ogamespec,
I'd like to write a little script for the sm83 leaflet map that automatically generates the tile images needed by the map from the source images, so that others can recreate the map themselves. I also want to properly document where the source images are from. Is the google drive the only source for the stitched together die shots? Or are they uploaded somewhere else?
I also used the topology you've drawn as an overlay, but I can only find it as JPG: https://github.com/emu-russia/dmgcpu/blob/main/imgstore/sm83/DMG01B_Core_Fused_Topo.jpg
I think I once extracted it as PNG (with transparent background) from a Photoshop file in the google drive. Could I use the PNGs in https://github.com/emu-russia/dmgcpu/tree/main/imgstore/topo as a new source? Do you remember in wich order I need to draw them over each other to generate a combined image?
It would be great to have this as a source, so that I can just regenerate the map tiles automatically whenever you make updates to it.

**Reply by ogamespec** [reply]
Hi!

Everything is here:
https://drive.google.com/drive/u/2/folders/1lEEsf0dfqN41lTlkm0sQyRvjxcGQjBN0

In the `Topo` folder there is a source file for Photoshop, which I used to research SM83. There are no other places :)

Regarding the masks, be careful. Due to microscope distortions, they may not exactly match the actual topology, for example, the one barely visible in the Silicon Pr0n pictures (I think they've changed their branding and are called something else now). Or in the SM83 picture from Gekkio. Especially in the area of ​​the flag setting logic – for some reason, everything turned out especially crooked there. :) It's best to use the slides separately, but they've gotten lost somewhere :( So you might have to tinker a bit, if it makes sense at all. :) Everything has already been studied down to the transistor.

I'll try to get the slides from the lap sessions, they're on my old nvme, which got damaged, I'm not sure I'll be able to extract them at all.

**Reply by msinger** [reply]
Okay, thanks. I'll try to extract everything from the newest Photoshop file.

**Reply by msinger** [reply]
I managed to automatically extract the die shots and the topology overlays from the psd file with ImageMagick. It works as long as the order of the layers in the file don't change. Here is the script in case you're interrested: https://github.com/msinger/sm83_map/blob/master/scripts/transform_die_shots.sh

**Reply by ogamespec** [reply]
It's a little off in the places I mentioned, but overall it looks nice :) Thanks! Descendants will appreciate it :)

<img width="1148" height="569" alt="image" src="https://github.com/user-attachments/assets/383ecd68-7334-48d3-bcc9-20be3d60a0cd" />

**Reply by ogamespec** [reply]
I think we just need to wait a little longer for neural networks to develop and then just give it two pictures with the prompt "Here are two pictures of the chip topology, align the second picture so that it matches the first" and it will straighten everything out itself :)

**Reply by msinger** [reply]
Yes, I think it is good enough. I actually find it impressive that those stitched together images even get so close, that the warping is so minimal.
